### PR TITLE
Added Podium feature

### PR DESCRIPTION
**Pull request by EPuncker**
Co-Authored-By: Zbizu <Zbizu@users.noreply.github.com>

### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
Added podium feature (item ids: 38629-38630), fully working except the outfit and mount name on description.
2021-12-16 23_10_52-Tibia - Perflex

credit: @Zbizu 

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

**Comment by Zbizu** [comment]
The podiums can be set from rme btw:
https://github.com/Zbizu/rme

Don't forget to build me an altar :P

**Review by DSpeichert — APPROVED** [review]

File: src/luascript.cpp [hunk 1]
```diff
@@ -7424,6 +7469,105 @@ int LuaScriptInterface::luaTeleportSetDestination(lua_State* L)
...
+int LuaScriptInterface::luaPodiumSetDirection(lua_State* L)
+{
+	// podium:setDirection(direction)
+	Podium* podium = getUserdata<Podium>(L, 1);
+	if (podium) {
+		podium->setDirection(getNumber<Direction>(L, 2));
+		g_game.updatePodium(podium);
+		pushBoolean(L, true);
+	}
+	else {
```
Comment (EPuncker) [inline src/luascript.cpp:7565]:
```suggestion
	} else {
```

File: src/luascript.cpp [hunk 2]
```diff
@@ -7424,6 +7469,105 @@ int LuaScriptInterface::luaTeleportSetDestination(lua_State* L)
...
+
+int LuaScriptInterface::luaPodiumSetDirection(lua_State* L)
+{
+	// podium:setDirection(direction)
+	Podium* podium = getUserdata<Podium>(L, 1);
+	if (podium) {
+		podium->setDirection(getNumber<Direction>(L, 2));
+		g_game.updatePodium(podium);
+		pushBoolean(L, true);
+	}
```
Comment (EPuncker) [inline src/luascript.cpp:7564]:
```suggestion
```

File: src/game.cpp [hunk 3]
```diff
@@ -5636,6 +5713,24 @@ void Game::removeBedSleeper(uint32_t guid)
 	}
 }
 
+void Game::updatePodium(Item* item) {
```
Comment (nekiro) [inline src/game.cpp:5716]: { in new line

File: src/item.h [hunk 4]
```diff
@@ -956,6 +963,10 @@ class Item : virtual public Thing
 			const ItemType& it = items[id];
 			return it.rotatable && it.rotateTo;
 		}
+		bool isPodium() const {
+			const ItemType& it = items[id];
+			return it.isPodium();
```
Comment (nekiro) [inline src/item.h:968]: one liner? items[id].isPodium()

File: src/podium.cpp [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+#include "otpch.h"
+
+#include "podium.h"
+#include "game.h"
+
+extern Game g_game;
+
+Attr_ReadValue Podium::readAttr(AttrTypes_t attr, PropStream& propStream)
+{
+	switch (attr) {
```
Comment (nekiro) [inline src/podium.cpp:29]: tabbing is off or is it github showing is bad?

File: src/podium.h [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+		void setOutfit(const Outfit_t& outfit) {
+			this->outfit = outfit;
+		}
+		const Outfit_t getOutfit() const {
+			return outfit;
+		}
+
+		bool hasFlag(PodiumFlags flag) const {
+			return (this->flags & flag) != 0;
```
Comment (nekiro) [inline src/podium.h:50]: use bitset for flags, everywhere

File: src/protocolgame.cpp [hunk 7]
```diff
@@ -3207,6 +3214,141 @@ void ProtocolGame::sendOutfitWindow()
 	writeToOutputBuffer(msg);
 }
 
+void ProtocolGame::sendPodiumWindow(const Item* item)
+{
+	if (!item) {
+		return;
```
Comment (nekiro) [inline src/protocolgame.cpp:3220]: tabs seem off again, github or broken tabs?

**Review by nekiro — CHANGES_REQUESTED** [review]

File: src/podium.h (hunk 6 above)
Comment (DSpeichert) [reply src/podium.h:50]: What about this @EPuncker ?

File: src/events.cpp [hunk 8]
```diff
@@ -868,6 +872,65 @@ void Events::eventPlayerOnTradeCompleted(Player* player, Player* target, Item* i
...
+
+	LuaScriptInterface::pushUserdata<Item>(L, item);
+	LuaScriptInterface::setItemMetatable(L, -1, item);
+
+	scriptInterface.callFunction(2);
+}
+
+void Events::eventPlayerOnPodiumEdit(Player* player, Item* item, const Outfit_t& outfit, bool podiumVisible, Direction direction)
+{
+	// Player:onPodiumEdit(item, outfit, direction, isVisible) or Player.onPodiumEdit(self, item, outfit, direction, isVisible)
```
Comment (DSpeichert) [inline src/events.cpp:904]:
```suggestion
	// Player:onPodiumEdit(item, outfit, direction, isVisible)
	// Player.onPodiumEdit(self, item, outfit, direction, isVisible)
```

**Review by DSpeichert — DISMISSED** [review]

**Pull request merged**